storage/innobase/xtrabackup/src/ddl_tracker.cc

```diff
@@ -306,9 +306,16 @@ static void copy_for_reduced(copy_thread_ctxt_t *ctxt) {
     if (ctxt->new_tables->find(node->space->id) == ctxt->new_tables->end()) {
       continue;
     }
-    std::string dest_name = node->name;
-    dest_name.append(EXT_NEW);
-    if (xtrabackup_copy_datafile_func(node, num, dest_name.c_str())) {
+    char dest_name[FN_REFLEN];
+    if (fsp_is_file_per_table(node->space->id, node->space->flags)) {
+      snprintf(dest_name, sizeof(dest_name), "%s%s%s",
+               fil_path_to_space_name(node->name), EXT_IBD.c_str(),
```

Comment:
Let's use a single place to convert a source filename to a destination filename.


Please move logic fil_path_to_space_name().

May be also also pass node->space->id or the entire node struct. use the space_id to determine if it is undo or not.

mysql> CREATE UNDO TABLESPACE UNDO_25 ADD DATAFILE 'blah_25.ibu';
Query OK, 0 rows affected (0.09 sec)

mysql> CREATE TABLESPACE ts1;
Query OK, 0 rows affected (0.01 sec)

datadir contents:
ls -l var/mysqld.1/data/
total 140612
-rw-rw---- 1 satya satya       56 Oct 10 11:02  auto.cnf
-rw-rw---- 1 satya satya   198343 Oct 10 11:03  binlog.000001
-rw-rw---- 1 satya satya       16 Oct 10 11:02  binlog.index
-rw-rw---- 1 satya satya 16777216 Oct 10 11:04  blah_25.ibu
-rw-rw---- 1 satya satya 16777216 Oct 10 11:03  bloh_25.ibu
-rw-rw---- 1 satya satya   114688 Oct 10 11:03  f3914a36-86ee-11ef-9c27-047bcbcb6b7e.ibd

Reply:
done